Fix graphicbuilder transform part 2

markschlosseratbentley opened the pull request:
The previous fix to the graphicbuilder transform problem actually resulted in some graphicbuilder primitives being transformed twice because they already had the transform applied to them. This fixes that issue by moving the transformation only to primitives that were missing it. This also adds code to ensure that GeometryAccumulator does not mutate the graphicbuilder's transform in-place.

pmconne commented on core/frontend/src/render/primitives/geometry/GeometryAccumulator.ts:
```diff
@@ -59,18 +59,20 @@ export class GeometryAccumulator {
-    if (this.haveTransform) this._transform.multiplyTransformTransform(transform, transform);
-    transform.multiplyRange(range, range);
+  private calculateTransform(transform: Transform, range: Range3d): Transform {
+    const xform = transform.clone();
```
Only clone if you have to mutate (`this.haveTransform`).

pmconne commented on core/frontend/src/render/primitives/geometry/GeometryPrimitives.ts:
```diff
@@ -270,6 +260,9 @@ class SolidPrimitiveGeometry extends Geometry {
-    return new PolyfacePrimitiveList(PolyfacePrimitive.create(this.displayParams, builder.claimPolyface()));
+    const pf = builder.claimPolyface();
+    if (!this.transform.isIdentity)
+      pf.tryTransformInPlace(this.transform);
```
Would be cheaper to transform the solid primitive before facetting than to transform the polyface.

pmconne commented:
Did you determine whether the test you added in previous PR would have succeeded even without your code changes?
If so, why did it continue to succeed with your code changes?
Do you need a better test?

markschlosseratbentley commented:
> Did you determine whether the test you added in previous PR would have succeeded even without your code changes?
> If so, why did it continue to succeed with your code changes?
> Do you need a better test?

@pmconne 
The test I added previously used an addShape decoration.  I incorrectly assumed that addShape decorations were not transforming properly like solidPrimitives.  Since the code change resulted in the transform being incorrectly applied twice to addShape decorations, the test "correctly" detected that the transformed decoration moved on the screen differently than before the code change.  My assumption of the problem was faulty.

I have added a test which actually tests the transform problem with a solidPrimitive.  This test correctly fails before this code change and succeeds after this code change.

pmconne commented on core/frontend/src/render/primitives/geometry/GeometryAccumulator.ts:
```diff
@@ -60,7 +60,7 @@ export class GeometryAccumulator {
 
   private calculateTransform(transform: Transform, range: Range3d): Transform {
-    const xform = transform.clone();
+    const xform = this.haveTransform ? transform : transform.clone();
```
Sorry to nitpick but
```
private calculateTransform(transform: Transform, range: Range3d): Transform {
  if (this.haveTransform)
    transform = this._transform.multiplyTransformTransform(transform);

  transform.multiplyRange(range, range);
  return transform;
}
```

markschlosseratbentley replied:
All good. That's a better usage of the API; thanks.
